asapdiscovery-ml/asapdiscovery/ml/schema_v2/trainer.py

```diff
@@ -0,0 +1,595 @@
+import json
+import pickle as pkl
+from pathlib import Path
+from time import time
+
+import numpy as np
+import torch
+import wandb
+from asapdiscovery.ml.es import BestEarlyStopping, ConvergedEarlyStopping
+from asapdiscovery.ml.schema_v2.config import (
+    DatasetConfig,
+    DatasetSplitterConfig,
+    EarlyStoppingConfig,
+    LossFunctionConfig,
+    OptimizerConfig,
+)
+from mtenn.config import ModelConfigBase
+from pydantic import BaseModel, Extra, Field, validator
+
+
+class Trainer(BaseModel):
```

Comment:
Reading through [this page](https://pytorch.org/docs/stable/notes/randomness.html) on reproducibility in pytorch, it looks like there are more sources of randomness than just the initial weights, and setting the model training process to be fully deterministic can slow it down. I think we have three options here:

1. Leave things as is, with random initial weights and non-deterministic training
2. Ensure we have identical weights when building a model from the same config, but accept the non-deterministic training
3. Make everything fully deterministic, potentially slowing stuff down

I feel like 2 is probably ok, but would love to hear your thoughts @hmacdope

Reply:
2. is perfectly fine, just wanted to check that weights came back ok.

Reply:
ok I have implemented 2 in `mtenn` with an optional random seed that, if given, will set the python, pytorch, and numpy random seeds before building the model. another question semi-related to all of this, does it make sense to also store the model weights in the config? that way a trained model can also be fully reproduced using just the config

Reply:
one more follow-up here: do you see any problem with using the same random seed for splitting the dataset and building the model? otherwise I'll need two separate random seed CLI args which I feel like gets a bit verbose